### PR TITLE
feat: add opentelemetry-resource-detector crate

### PR DESCRIPTION
**Pull request by rogercoll**
Moves the implementation of the OS and Process resource detectors from the [opentelemetry-sdk](https://github.com/open-telemetry/opentelemetry-rust/tree/main/opentelemetry-sdk/src/resource) to its own contrib crate. The reason behind it is due to their experimental nature in semantic conventions:
- OS: https://github.com/open-telemetry/opentelemetry-specification/blob/v1.6.x/specification/resource/semantic_conventions/os.md
- Process: https://github.com/open-telemetry/opentelemetry-specification/blob/v1.6.x/specification/resource/semantic_conventions/process.md

Discussion: https://github.com/open-telemetry/opentelemetry-rust/pull/1610

## Changes

Adds a new crate for contrib resource detectors.

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-rust/blob/main/CONTRIBUTING.md) guidelines followed
* [x] Unit tests added/updated (if applicable)
* [x] Appropriate `CHANGELOG.md` files updated for non-trivial, user-facing changes
* [x] Changes in public API reviewed (if applicable)

### CODE REVIEW

File: opentelemetry-resource-detector/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+[package]
+name = "opentelemetry-resource-detector"
```
Comment (cijothomas) [inline opentelemetry-resource-detector/Cargo.toml:2]:
```suggestion
name = "opentelemetry-resource-detectors"
```

Would it make sense to use plural name, given this has multiple detectors!

Comment (rogercoll) [reply opentelemetry-resource-detector/Cargo.toml]: Make sense to me, changed in https://github.com/open-telemetry/opentelemetry-rust-contrib/pull/49/commits/380ca691c838d27667c03df30b5d68f0d8b72596

File: opentelemetry-resource-detectors/README.md [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+[splash]: https://raw.githubusercontent.com/open-telemetry/opentelemetry-rust/main/assets/logo-text.png
+
+# OpenTelemetry Contrib Resource Detectors
+
+Community supported Resource detectors implementations for applications instrumented with [`OpenTelemetry`].
+
+## Overview
+
+[`The contrib resource detector`] crate provides a collection of tools to detect resources from the runtime. This crate provides additional detectors for OS and process-related resources. Despite not being included in the SDK due to their lack of stability, these detectors can facilitate gathering additional telemetry information.
```
Comment (lalitb) [inline opentelemetry-resource-detectors/README.md:11]: nit - `opentelemetry-resource-detectors` crate

File: opentelemetry-resource-detectors/README.md [hunk 3]
```diff
@@ -0,0 +1,18 @@
+![OpenTelemetry — An observability framework for cloud-native software.][splash]
+
+[splash]: https://raw.githubusercontent.com/open-telemetry/opentelemetry-rust/main/assets/logo-text.png
+
+# OpenTelemetry Contrib Resource Detectors
```
Comment (lalitb) [inline opentelemetry-resource-detectors/README.md:5]: nit - remove contrib ?

File: opentelemetry-resource-detectors/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,18 @@
+[package]
+name = "opentelemetry-resource-detectors"
+version = "0.1.0"
+edition = "2021"
+description = "OpenTelemetry contrib resource detectors"
```
[the remark this answers is not in the document]
Comment (cijothomas) [reply opentelemetry-resource-detectors/Cargo.toml:5]: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/tree/main/src OTel .NET chose to keep separate package for each detector.. Any preference for us here? I am okay to have a single detectors package supporting the common ones (i.e this PR), and they vendor specific ones like AWS/Azure/etc.c an be own crates.

[the remark this answers is not in the document]
Comment (lalitb) [reply opentelemetry-resource-detectors/Cargo.toml:5]: yeah, I was also thinking on that. So this package will have all detectors from (say)- https://github.com/open-telemetry/semantic-conventions/tree/main/model/resource, except the vendor ones ?

File: opentelemetry-resource-detectors/README.md (hunk 3 above)
Comment (rogercoll) [reply opentelemetry-resource-detectors/README.md]: Fixed in https://github.com/open-telemetry/opentelemetry-rust-contrib/pull/49/commits/98052034707af7285864ea267c899b9b8b632e08
Thank you

File: opentelemetry-resource-detectors/README.md (hunk 2 above)
Comment (rogercoll) [reply opentelemetry-resource-detectors/README.md]: Fixed in https://github.com/open-telemetry/opentelemetry-rust-contrib/pull/49/commits/98052034707af7285864ea267c899b9b8b632e08
Thank you

File: opentelemetry-resource-detectors/Cargo.toml (hunk 4 above)
[the remark this answers is not in the document]
Comment (rogercoll) [reply opentelemetry-resource-detectors/Cargo.toml]: I would suggest having a single package for now and use feature flags for each detector subset: `system, os, cloud, cloud/aws`. This way, we can avoid importing unneeded dependencies. Does .NET have something similar like Rust feature flags?

Description nit applied in https://github.com/open-telemetry/opentelemetry-rust-contrib/commit/98052034707af7285864ea267c899b9b8b632e08

**Review by lalitb — APPROVED** [review]
Thanks.

File: opentelemetry-resource-detectors/src/process.rs [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+            ),
+            KeyValue::new(
+                opentelemetry_semantic_conventions::resource::PROCESS_PID,
+                id() as i64,
+            ),
+        ])
+    }
+}
+
+#[cfg(target_os = "linux")]
```
Comment (cijothomas) [inline opentelemetry-resource-detectors/src/process.rs:41]: nothing linux specific right?

**Review by cijothomas — APPROVED** [review]

File: opentelemetry-resource-detectors/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,18 @@
+[package]
+name = "opentelemetry-resource-detectors"
+version = "0.1.0"
+edition = "2021"
+description = "A collection of community supported resource detectors for OpenTelemetry"
+homepage = "https://github.com/open-telemetry/opentelemetry-rust-contrib/tree/main/opentelemetry-resource-detectors"
+repository = "https://github.com/open-telemetry/opentelemetry-rust-contrib/tree/main/opentelemetry-resource-detectors"
+readme = "README.md"
+keywords = ["opentelemetry", "resource", "detector"]
```
Comment (TommyCpp) [inline opentelemetry-resource-detectors/Cargo.toml:9]: Not sure if `resource` and `detector` are best keywords. Maybe `tracing`?

**Pull request merged**